clippy_utils/src/eager_or_lazy.rs

```diff
@@ -207,6 +214,37 @@ fn expr_eagerness<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) -> EagernessS
                         self.cx.typeck_results().expr_ty(e).kind(),
                         ty::Bool | ty::Int(_) | ty::Uint(_),
                     ) => {},
+
+                // `>>` and `<<` panic when the right-hand side is greater than or equal to the number of bits in the
+                // type of the left-hand side, or is negative. Doesn't need `constant` on the whole expression
+                // because only the right side matters.
+                ExprKind::Binary(op, left, right)
+                    if matches!(op.node, BinOpKind::Shl | BinOpKind::Shr)
+                        && let left_ty = self.cx.typeck_results().expr_ty(left)
+                        && let left_bits = left_ty.int_size_and_signed(self.cx.tcx).0.bits()
+                        && constant(self.cx, self.cx.typeck_results(), right)
+                            .and_then(|c| c.int_value(self.cx, left_ty))
+                            .map_or(true, |c| match c {
+                                FullInt::S(i) => i >= i128::from(left_bits) || i.is_negative(),
+                                FullInt::U(i) => i >= u128::from(left_bits),
+                            }) =>
```

Comment:
Just FYI. The compiler will already check for this.

Reply:
Can you elaborate? I'm not sure I understand what needs to be changed? Do you mean that rustc's `arithmetic_overflow` lint detects overflowing exprs, so it doesn't matter if we end up linting because the user gets warned anyway when the expr is a constant, so we just need to check for `constant(right).is_none()`?

Reply:
Since the code as written is almost certainly has the wrong shift amount, it would be better to suggest eagerly evaluating to avoid having the lint only come up after having fixed the first problem.

e.g. When changing `x >> 32` to `x >> 31` it would have been better to get both lints up front.

Reply:
Do you think we should do the same for `/` and `%`, then? That is, still lint for something like `x / 0`. That would follow the same logic here afaict?

Reply:
@Jarcho ping? :)

Reply:
That sounds reasonable.

Reply:
I was going to do it by checking if `constant(right).is_none()` because only then could it be dependent on runtime values. The compiler will catch it for `Some(_)` constants and we can still early-suggest it like you said because we'd only delay the lint by not immediately suggesting eager evaluation.

I guess if we stay consistent with this, we really should be doing it for all arithmetic operations then, i.e., `u32::MAX + 1` could still be evaluated eagerly. It's a programmer error and the compiler emits an error.

One (more or less -- it's only a FN) problem with this is that something like `true.then(|| 1 >> (2 >> 32))` wouldn't lint, because the RHS of the outer right shift will evaluate to `None` in clippy's const eval.

So in a sense we'd need to know if the failure case is because an error in const eval happened (overflow), or the user tried to do non-const things, such as referring to variables, like `1 >> n`. Only in the latter case do we want to avoid linting.

`is_const_evaluatable` has the problem I mentioned earlier (it can be dependent on downstream consts). There's not really any better util for this, right? I guess we have to either make clippy's const eval return why it failed, or accept the false negative, or come up with some other way.

Reply:
Maybe we could run `constant()` on the leaves of nested binary expressions, e.g., for `((1 + (n + 3)) >> (5 + (6 * 7))) >> (5 + (6 * 8))` that would be `n`, `3`, `6`, `7`, `6`, `8`. If any of them return `None` in const eval, we should not lint.

If all of them return `Some(_)`, the compiler should be able to either verify that it's okay or it will emit an error that it overflows at runtime. We can still emit a warning in any case.

It's possible that I'm overcomplicating this though ^^

Reply:
I personally don't think it's worth the effort to propagate the overflow since it's a thing that would be fixed after seeing the error anyways. If you want to code it then go for it. It does make things slightly nicer.

Evaluating the children of a node that failed to evaluate is definitely the wrong way to handle this.

Reply:
> Evaluating the children of a node that failed to evaluate is definitely the wrong way to handle this.

Why? I'm not seeing where the approach that I mentioned in my last comment would break and I don't think that would add a ton of complexity. That is, checking if all leaves are able to evaluate in `constant()` and if they aren't, overflow can be ruled out and it must mean that it's runtime dependent.

I agree though, this doesn't really matter all that much. Rustc is going to emit an error for overflow and the lint would trigger later after that's fixed no matter if there are FNs here or not.

Reply:
It would work, but propagating the overflow error back up is both the better solution and not any extra work to code.

Reply:
Yeah, ok, makes sense now. I feel like that'd involve a lot of changes within, switching from `Option` to `Result`, to propagate it. That could probably be a PR on its own. Let's just stick to only eagerly linting overflow at the root binary expression like `1 >> 32` for now. This all seems like a nice-to-have, not strictly required to fix FNs or FPs.